### ISSUE TITLE
Legend rendering glitch on Canvas

### ISSUE THREAD

**Issue by arvind** [vega/vega#694]
With the attached spec, drag to brush, and then drag the brush back and forth over the legend to observe a glitch. This only occurs with the Canvas renderer; the legend stays put with SVG. Perhaps something in bounds calculation is going awry? 

![untitled2](https://cloud.githubusercontent.com/assets/42262/21561523/7a7c853e-ce93-11e6-8f31-a08de1633b9a.gif)

```json
{
  "autosize": "pad",
  "data": [
    {"name": "source","url": "data/cars.json"},
    {"name": "id_store"}
  ],
  "marks": [
    {
      "type": "rect",
      "encode": {
        "enter": {"fill": {"value": "#eee"}},
        "update": {
          "x": {"scale": "x","signal": "id[0].extent[0]"},
          "x2": {"scale": "x","signal": "id[0].extent[1]"},
          "y": {"scale": "y","signal": "id[1].extent[0]"},
          "y2": {"scale": "y","signal": "id[1].extent[1]"}
        }
      }
    },
    {
      "name": "marks",
      "type": "symbol",
      "from": {"data": "source"},
      "encode": {
        "update": {
          "x": {"scale": "x","field": "Horsepower"},
          "y": {"scale": "y","field": "Miles_per_Gallon"},
          "size": {"value": 30},
          "shape": {"value": "circle"},
          "strokeWidth": {"value": 2},
          "opacity": {"value": 0.7},
          "stroke": {"scale": "color","field": "Origin"},
          "fill": {"value": "transparent"}
        }
      }
    },
    {
      "name": "id_brush",
      "type": "rect",
      "encode": {
        "enter": {"fill": {"value": "transparent"}},
        "update": {
          "x": {"scale": "x","signal": "id[0].extent[0]"},
          "x2": {"scale": "x","signal": "id[0].extent[1]"},
          "y": {"scale": "y","signal": "id[1].extent[0]"},
          "y2": {"scale": "y","signal": "id[1].extent[1]"}
        }
      }
    }
  ],
  "scales": [
    {
      "name": "x",
      "type": "linear",
      "range": [0,200],
      "domain": {"data": "source","field": "Horsepower"}
    },
    {
      "name": "y",
      "type": "linear",
      "range": [200,0],
      "domain": {"data": "source","field": "Miles_per_Gallon"}
    },
    {
      "name": "color",
      "type": "ordinal",
      "scheme": "category10",
      "domain": {"data": "source","field": "Origin","sort": true}
    }
  ],
  "legends": [{"stroke": "color"}],
  "signals": [
    {
      "name": "unit",
      "value": {},
      "on": [{"events": "mousemove","update": "group()"}]
    },
    {"name": "width","value": 200},
    {"name": "height","value": 200},
    {
      "name": "id_x",
      "value": [],
      "on": [
        {
          "events": {
            "source": "view",
            "type": "mousedown",
            "filter": [
              "!event.item || (event.item && event.item.mark.name !== \"id_brush\")"
            ]
          },
          "update": "invert(\"x\", [x(unit), x(unit)])"
        },
        {
          "events": {
            "source": "window",
            "type": "mousemove",
            "between": [
              {
                "source": "view",
                "type": "mousedown",
                "filter": [
                  "!event.item || (event.item && event.item.mark.name !== \"id_brush\")"
                ]
              },
              {"source": "window","type": "mouseup"}
            ]
          },
          "update": "[id_x[0], invert(\"x\", clamp(x(unit), 0, width))]"
        },
        {
          "events": {"signal": "id_translate_delta"},
          "update": "[id_translate_anchor.extent_x[0] + span(id_translate_anchor.extent_x) * id_translate_delta.x / id_translate_anchor.width, id_translate_anchor.extent_x[1] + span(id_translate_anchor.extent_x) * id_translate_delta.x / id_translate_anchor.width]"
        }
      ]
    },
    {
      "name": "id_y",
      "value": [],
      "on": [
        {
          "events": {
            "source": "view",
            "type": "mousedown",
            "filter": [
              "!event.item || (event.item && event.item.mark.name !== \"id_brush\")"
            ]
          },
          "update": "invert(\"y\", [y(unit), y(unit)])"
        },
        {
          "events": {
            "source": "window",
            "type": "mousemove",
            "between": [
              {
                "source": "view",
                "type": "mousedown",
                "filter": [
                  "!event.item || (event.item && event.item.mark.name !== \"id_brush\")"
                ]
              },
              {"source": "window","type": "mouseup"}
            ]
          },
          "update": "[id_y[0], invert(\"y\", clamp(y(unit), 0, height))]"
        },
        {
          "events": {"signal": "id_translate_delta"},
          "update": "[id_translate_anchor.extent_y[0] + span(id_translate_anchor.extent_y) * id_translate_delta.y / id_translate_anchor.height, id_translate_anchor.extent_y[1] + span(id_translate_anchor.extent_y) * id_translate_delta.y / id_translate_anchor.height]"
        }
      ]
    },
    {
      "name": "id_size",
      "value": [],
      "on": [
        {
          "events": {
            "source": "view",
            "type": "mousedown",
            "filter": [
              "!event.item || (event.item && event.item.mark.name !== \"id_brush\")"
            ]
          },
          "update": "{x: x(unit), y: y(unit), width: 0, height: 0}"
        },
        {
          "events": {
            "source": "window",
            "type": "mousemove",
            "between": [
              {
                "source": "view",
                "type": "mousedown",
                "filter": [
                  "!event.item || (event.item && event.item.mark.name !== \"id_brush\")"
                ]
              },
              {"source": "window","type": "mouseup"}
            ]
          },
          "update": "{x: id_size.x, y: id_size.y, width: abs(x(unit) - id_size.x), height: abs(y(unit) - id_size.y)}"
        }
      ]
    },
    {
      "name": "id",
      "update": "[{field: \"Horsepower\", extent: id_x}, {field: \"Miles_per_Gallon\", extent: id_y}]"
    },
    {
      "name": "id_translate_anchor",
      "value": {},
      "on": [
        {
          "events": "@id_brush:mousedown",
          "update": "{x: x(unit), y: y(unit), width: id_size.width, height: id_size.height, extent_x: slice(id_x), extent_y: slice(id_y), }"
        }
      ]
    },
    {
      "name": "id_translate_delta",
      "value": {},
      "on": [
        {
          "events": "[@id_brush:mousedown, window:mouseup] > window:mousemove",
          "update": "{x: x(unit) - id_translate_anchor.x, y: y(unit) - id_translate_anchor.y}"
        }
      ]
    },
    {
      "name": "id_tuple",
      "on": [
        {
          "events": {"signal": "id"},
          "update": "{unit: unit.datum && unit.datum._id, intervals: id}"
        }
      ]
    },
    {
      "name": "id_modify",
      "on": [
        {
          "events": {"signal": "id"},
          "update": "modify(\"id_store\", id_tuple, {unit: id_tuple.unit})"
        }
      ]
    }
  ]
}```

**Comment by jheer** [comment]
Thanks, will look into this soon. This may be associated with the `ViewLayout` including the brush element within a bounds calculation. Will investigate alongside #691.

**Comment by jheer** [comment]
I have tracked the behavior to the `ViewLayout` operator in vega-view. The brush bounds were being included in the layout calculation, causing the legend to be "pushed" aside by the brush.

In response, I've updated the legend layout to only consider axis bounds and sub-group bounds (in the case a legend is being shown for a collection of plots, as in the SPLOM and barley examples). In the `ViewLayout` code, this is accomplished by using the `viewBounds`, not the `markBounds`, as part of the legend layout calculation; see vega/vega-view@1e4b8b1.

We now ignore marks at the same plotting "level"; for example, a large circle on the border of a scatter plot will no longer cause the legend to "push" to the side. While this introduces the potential for some occlusion, it also makes the legend layout more stable and predictable. Designers can always increase the legend padding config to deal with large marks that exceed the normal width/height of a plot.
